.github/workflows/publish_charm.yaml

```diff
@@ -63,6 +63,53 @@ jobs:
           docker pull $image
           imageId=$(docker images $image --format "{{.ID}}")
           charmcraft upload-resource $charmName ${{ matrix.image }}-image --image=$imageId --verbosity=trace
+  lookup-external-images:
+    name: Lookup External Images
+    runs-on: ubuntu:latest
+    outputs:
+      images: ${{ env.IMAGES }}
+    steps:
+      - uses: actions/checkout@v3
+      - if: inputs.working-directory != './'
+        name: Change directory
+        run: |
+          TEMP_DIR=$(mktemp -d)
+          cp -rp ${{inputs.working-directory}}/. $TEMP_DIR
+          rm -rf .* * || :
+          cp -rp $TEMP_DIR/. .
+          rm -rf $TEMP_DIR
+      - name: Extract External Images
+        run: |
+          IMAGES=$(yq -c '.resources | to_entries | map(select(.value.type == "oci-image" and .value."upstream-source")) | map(.key)' metadata.yaml)
+          echo "External Images: $IMAGES"
+          echo "IMAGES=$IMAGES" >> $GITHUB_ENV
+  publish-external-images:
+    name: Push External Images to Charmhub
+    runs-on: ubuntu:latest
+    needs: [ lookup-external-images ]
+    if: ${{ needs.lookup-external-images.outputs.images != '[]' }}
+    strategy:
+      matrix:
+        image: ${{ fromJSON(needs.lookup-external-images.outputs.images) }}
+    steps:
+      - uses: actions/checkout@v3
+      - if: inputs.working-directory != './'
+        name: Change directory
+        run: |
+          TEMP_DIR=$(mktemp -d)
+          cp -rp ${{inputs.working-directory}}/. $TEMP_DIR
+          rm -rf .* * || :
+          cp -rp $TEMP_DIR/. .
+          rm -rf $TEMP_DIR
+      - name: Publish image
+        env:
+          CHARMCRAFT_AUTH:  ${{ secrets.CHARMHUB_TOKEN }}
+        run: |
+          image=${{ matrix.image }}
+          charmName=$(yq '.name' metadata.yaml)
+          docker pull $image
+          imageId=$(docker images $image --format "{{.ID}}")
+          charmcraft upload-resource $charmName ${{ matrix.image }}-image --image=$imageId --verbosity=trace
```

Comment:
I'm not familiar if there is some kind of mechanism to prevent that: usually, external images are always the same. Will the upload recognize that or every time will create a new revision even if there is the same image already uploaded?

Reply:
I just tested on my local machine, CharmHub will create a new revision for the same image.
But I think this is okay since the image is stored by sha256 key, so the duplicate image will not cause more storage. Revision is just an entry in the database.

Reply:
And there's anyway no way easy way to get the checksump of already uploaded OCI images to avoid reuploading the same image.

Do we want to keep the `--verbosity=trace` ?